change type from eigen to vector

bbbales2 opened the pull request:
Don't merge this, I just opened it so we can easily see the changes on your branch vs. your develop

Commit:
change type from eigen to vector

bbbales2 commented:
Very cool! I will look at this tomorrow!

Commit:
upload stan file
wq
exit
exit()
kill
intr
wq

Commit:
Revert "upload stan file"

This reverts commit c30b1b83f1bb2ce835a77c9a4792d7de7ac8e0b6.

Commit:
test pass version allocate memory for yVals

Commit:
unpassed sorted version

bbbales2 commented on stan/math/prim/fun/interp1.hpp:
```diff
@@ -1,45 +1,48 @@
+                xData[i] = vec[i].first;
+                yData[i] = vec[i].second;
+            }
+            for (int i = 0; i < size_xData; i++){
```
There's an off by one error here. So since xData is size `size_xData` (I find it convenient to use single capital letters like `N` or `M` for these sorts of things), this loop should be:

```
for (int i = 0; i < size_xData - 1; i++)
```

to avoid overflow.

bbbales2 commented on stan/math/prim/fun/interp1.hpp:
```diff
@@ -1,45 +1,48 @@
-            } else {
-                while (x > xData[i + 1])
-                    i++;                                   // find left end of interval for interpolation
+            // sort if xData is not monotone
```
It's correct to try to sort this sort of thing, but that's gonna be kindof expensive.

It will be more efficient if we ask the user to provide an array that is already sorted.

We should also make sure that xData and yData are the same size:

So you can do that with code like this:
```
check_size_match("interp1", "x", xDataPre.size(), "y", yDataPre.size());
check_sorted("interp1", "x", xDataPre);
```

And then we can assume that the data has already been sorted to work here (presumably someone could do that in R or somewhere else and everything will just end up running faster)

bbbales2 commented on stan/math/prim/fun/interp1.hpp:
```diff
@@ -1,45 +1,48 @@
-        Eigen::VectorXd yVals;
-        int size_x = xData.size();
-        int size_xData = xData.size();
+        template <typename T>
```
I see you're already templating the type! That is good :D.

Start with only templating xTest though. Assume `xData` and `yData` are `std::vector<doubles>`.

Whenever you do calculations with multiple types, the output type is a function of all the different input types. I think Bob kinda talks about this like a least upper bound or something?

Like, if you have integers and doubles, and you do math, in general the output will be a double to not lose precision (cuz 1.5 isn't an integer, but 3 / 2.0 is 1.5).

Similarly if you have a float and a double and do math, in general the output should be a double again to not lose precision (the double is higher precision).

So in Stan if you do Math on a double and an autodiff type (which is why we template stuff), the output will be that autodiff type.

So we use the `return_type_t` metaprogram in Stan (this is specific for Stan) to compute return types.

So if you had the fully templated function we write that like this:

```cpp
template<typename T1, typename T2, typename T3>
inline std::vector<return_type_t<T1, T2, T3>>
interp1(const std::vector<T1>& xDataPre,
  const std::vector<T2>& yDataPre,
  const std::vector<T3>& xTest) {
  ...
}
```

But for now just start with only `xTest` templated:
```
template<typename T>
inline std::vector<T>
interp1(const std::vector<double>& xDataPre,
  const std::vector<double>& yDataPre,
  const std::vector<T>& xTest) {
  ...
}
```

This means all the temporaries you compute in Stan that are a function of a `T` should themselves have the `T` type.

Edit: updated a type to be more specific

hyunjimoon replied:
Thanks for the explanation; have changed xTest only <T> in the new version. Could I ask what could go wrong with the first(original) one please? 

```
template<typename T>
inline std::vector<double> interp1(const std::vector<T>& xDataPre,
  const std::vector<T>& yDataPre,
  const std::vector<T>& xTest) {
  ...
}
```

```
template<typename T1, typename T2, typename T3>
inline std::vector<return_type_t<T1, T2, T3>>
interp1(const std::vector<T1>& xDataPre,
  const std::vector<T2>& yDataPre,
  const std::vector<T3>& xTest) {
  ...
}
```

bbbales2 replied:
If all the template types are the same, then all three inputs need to be autodiff types.

In Stan, the functions need to be a mix of doubles or autodiff types. In this case there are 3 parameters, so 3 things that can either be doubles or autodiff types -> 8 total possible functions.

bbbales2 commented on stan/math/prim/fun/sort_asc.hpp:
```diff
@@ -18,12 +18,12 @@ namespace math {
-  return xs;
-}
+        template <typename T>
+        inline std::vector<T> sort_asc(std::vector<T> xs) {
```
Oh yeah set your editor to 2 space indents. There's an autoformatter for stuff going into Stan but it's just more space efficient :D

hyunjimoon replied:
done!

bbbales2 reviewed (commented):
I made this comments somewhere else, but I'm copying them here so they're easier to find (they'll be attached specifically to this pull).

Commit:
update template type etc

bbbales2 commented on stan/math/prim/fun/interp1.hpp:
```diff
@@ -1,45 +1,36 @@
+        template <typename T>
+        inline std::vector<double> interp1(const std::vector<double>& xData, const std::vector<double>& yData, const std::vector<T>& xTest)
+        {
+            using stan::math::index_type_t;
```
Since we're not making assumptions about the input we need to check those assumptions:

```
check_size_match("interp1", "x", xDataPre.size(), "y", yDataPre.size());
check_sorted("interp1", "x", xDataPre);
```

hyunjimoon replied:
done

bbbales2 commented on test/unit/math/prim/fun/interp1_test.dSYM/Contents/Info.plist:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Delete this file and also `test/unit/math/prim/fun/interp1_test.dSYM/Contents/Resources/DWARF/interp1_test`.

It's worth double checking what you're committing, since once it gets committed, it stays in the repo history even if we revert it.

hyunjimoon replied:
ok, I will double check

bbbales2 commented on stan/math/prim/fun/interp1.hpp:
```diff
@@ -1,45 +1,36 @@
-        int size_x = xData.size();
-        int size_xData = xData.size();
+        template <typename T>
+        inline std::vector<double> interp1(const std::vector<double>& xData, const std::vector<double>& yData, const std::vector<T>& xTest)
```
The next step here is getting gradients.

So back to the autodiff comment, but the output type scalar type should be the least-upper bound of the input scalar types (I think LUB of the scalar types is the way to talk about this, anyway, could be wrong).

Since you have a `T`, we have to think about this.

In Stan the `T` might be instantiated to `double`, `var`, `fvar<double>`, or some other higher order autodiff type.

If the input is an autodiff type, then the output will be an autodiff type too. So the output type in this case should be `std::vector<stan::math::return_type_t<double, double, T>>` which just simplifies to `std::vector<T>` if `T` is any of the types we might expect.

This also means the temporaries in this calculation should be `T`.

bbbales2 replied:
Add this code as the file `test/unit/math/rev/fun/interp1_test.cpp`:

```
#include <stan/math/rev.hpp>
#include <gtest/gtest.h>
#include <limits>

TEST(AgradRev, interp1_gradient) {
  using stan::math::var;
  using stan::math::interp1;

  std::vector<double> x = { 0.1, 1.2 };
  std::vector<double> y = { 1.0, -2.0 };

  var x_int = 0.7;
  std::vector<var> x_int_vec = { x_int };

  std::vector<var> y_int_vec = interp1(x, y, x_int_vec);
  var y_int = y_int_vec[0];

  y_int.grad();

  // x_int.adj() should contain dy_int/dx_int                                                                                                                                
  EXPECT_FLOAT_EQ(-3.0 / 1.1, x_int.adj());
}
```

I tried to push to the repo but it said I didn't have write access. Can you give me write access? It'll be easier if I can push directly.

bbbales2 reviewed (commented):
I left a couple comments. Time for testing the autodiff. I will upload a test file shortly.

bbbales2 commented:
Once we have reverse mode working and everything tested we need to:

1. Install a cmdstanr using this branch with Rok's script https://github.com/stan-dev/cmdstanr/issues/300#issuecomment-702824987
2. Set a cmdstan model up to use the function we wrote and verify it works
3. Write up the whole process

I think that's it, at least. There's probably a pile of details I'm leaving out.

Commit:
autodiff test and change temporaries to type T

hyunjimoon committed:
Update if_else_test.cpp

sorry for this typo; do you use revert in this case? and if I have questions would it be wise to make this as a pull request?

hyunjimoon committed:
Merge pull request #2 from hyunjimoon/hyunjimoon-patch-1

typo

hyunjimoon commented:
Regarding 
2. Set a cmdstan model up to use the function we wrote and verify it works

How do I expose the function to stan? The following is the `install.sh` file I used last summer. Is it the same process?

```
#! /bin/bash
git clone https://github.com/stan-dev/stanc3.git
cd stanc3
git checkout try-laplace_approximation
cd ..
git clone https://github.com/stan-dev/cmdstan.git
cd cmdstan
make stan-update
cd stan
git checkout feature/issue-2522-laplace
cd lib/stan_math
git checkout feature/issue-755-laplace
cd ..
cd ..
cd ..
# make build
STANC3=../stanc3 make examples/bernoulli/bernoulli
```

bbbales2 commented:
Oh nah we don't have to add the function to stanc3. There is a workaround we should use.

So there is a flag we can pass to the stanc3 compiler, `--allow-undefined`, that will allow us to use undefined functions in Stan.

SO we can just add:

```
functions {
  real[] interp1(real[] x, real[] y, real[] x_test);
}
```

And add the `--allow-undefined` flag and the model will build.

I think you can add `stanc_options = list("--allow-undefined")` to compile options in cmdstanr: `https://mc-stan.org/cmdstanr/reference/model-method-compile.html`

So try installing a custom cmdstan via Roks install script: https://github.com/stan-dev/cmdstanr/issues/300#issuecomment-702824987 and point the math_url and math_branch at this github (`https://github.com/hyunjimoon/math`) and this branch (`feature/interp1`) and see if it installs.

A full test model:
```stan
functions {
  real[] interp1(real[] x, real[] y, real[] x_test);
}

transformed data {
  real x[2] = { -5.0, 10.0 };
  real y[2] = { -3.0, 0.0 };
}

parameters {
  real<lower = -0.5, upper = 10.0> x_int[1];
}

transformed parameters {
  real y_int = interp1(x, y, x_int)[1];
}

model {
  x_int ~ normal(0, 5);
}
```

Because the slope of the line here is 1/5, then `y_int` should have a standard normal distribution if everything is working fine (so just check that the posterior intervals line up with what you'd expect from a standard normal, like `qnorm(c(0.05, 0.25, 0.75, 0.95), 0, 1)` or something)

hyunjimoon committed:
add pstream argument

hyunjimoon committed:
change every input type to template

hyunjimoon committed:
Update interp1_test.cpp

hyunjimoon committed:
update type of linear slope 

@bbbales2 I cannot understand `std::vector<double> dydx(N);` could have passed the prim test; is dydx not used in prim? This type error was detected only in rev mode test.

hyunjimoon commented:
`requires at most 2 arguments, but 3 were provided`? is this the cause? 
error log from running this [script](https://github.com/hyunjimoon/customFunc/blob/master/interp1.R).
```
> mod <- cmdstan_model("interp1.stan", stanc_options = list("allow-undefined" = TRUE))
Compiling Stan program...
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:182:15: error: call to 'interp1' is ambiguous
      y_int = interp1(x, y, x_int, pstream__)[(1 - 1)];
              ^~~~~~~
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:392:14: note: in instantiation of function template specialization 'interp1_model_namespace::interp1_model::log_prob<false, false, stan::math::var_value<double> >' requested here
      return log_prob<propto__,jacobian__,T_>(vec_params_r, vec_params_i, pstream);
             ^
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/src/stan/model/model_base_crtp.hpp:95:50: note: in instantiation of function template specialization 'interp1_model_namespace::interp1_model::log_prob<false, false, stan::math::var_value<double> >' requested here
    return static_cast<const M*>(this)->template log_prob<false, false>(theta,
                                                 ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:82:3: note: in instantiation of member function 'stan::model::model_base_crtp<interp1_model_namespace::interp1_model>::log_prob' requested here
  ~interp1_model() final { }
  ^
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/lib/stan_math/stan/math/prim/fun/interp1.hpp:8:55: note: candidate function [with T1 = double, T2 = double, T3 = stan::math::var_value<double>]
        inline std::vector<return_type_t<T1, T2, T3>> interp1(const std::vector<T1>& xData, const std::vector<T2>& yData, const std::vector<T3>& xTest, std::ostream* pstream__ = nullptr)
                                                      ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:72:1: note: candidate function [with T0__ = double, T1__ = double, T2__ = stan::math::var_value<double>]
interp1(const std::vector<T0__>& x, const std::vector<T1__>& y,
^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:182:15: error: call to 'interp1' is ambiguous
      y_int = interp1(x, y, x_int, pstream__)[(1 - 1)];
              ^~~~~~~
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:392:14: note: in instantiation of function template specialization 'interp1_model_namespace::interp1_model::log_prob<false, true, stan::math::var_value<double> >' requested here
      return log_prob<propto__,jacobian__,T_>(vec_params_r, vec_params_i, pstream);
             ^
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/src/stan/model/model_base_crtp.hpp:106:50: note: in instantiation of function template specialization 'interp1_model_namespace::interp1_model::log_prob<false, true, stan::math::var_value<double> >' requested here
    return static_cast<const M*>(this)->template log_prob<false, true>(theta,
                                                 ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:82:3: note: in instantiation of member function 'stan::model::model_base_crtp<interp1_model_namespace::interp1_model>::log_prob_jacobian' requested here
  ~interp1_model() final { }
  ^
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/lib/stan_math/stan/math/prim/fun/interp1.hpp:8:55: note: candidate function [with T1 = double, T2 = double, T3 = stan::math::var_value<double>]
        inline std::vector<return_type_t<T1, T2, T3>> interp1(const std::vector<T1>& xData, const std::vector<T2>& yData, const std::vector<T3>& xTest, std::ostream* pstream__ = nullptr)
                                                      ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:72:1: note: candidate function [with T0__ = double, T1__ = double, T2__ = stan::math::var_value<double>]
interp1(const std::vector<T0__>& x, const std::vector<T1__>& y,
^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:182:15: error: call to 'interp1' is ambiguous
      y_int = interp1(x, y, x_int, pstream__)[(1 - 1)];
              ^~~~~~~
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:392:14: note: in instantiation of function template specialization 'interp1_model_namespace::interp1_model::log_prob<true, false, stan::math::var_value<double> >' requested here
      return log_prob<propto__,jacobian__,T_>(vec_params_r, vec_params_i, pstream);
             ^
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/src/stan/model/model_base_crtp.hpp:117:50: note: in instantiation of function template specialization 'interp1_model_namespace::interp1_model::log_prob<true, false, stan::math::var_value<double> >' requested here
    return static_cast<const M*>(this)->template log_prob<true, false>(theta,
                                                 ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:82:3: note: in instantiation of member function 'stan::model::model_base_crtp<interp1_model_namespace::interp1_model>::log_prob_propto' requested here
  ~interp1_model() final { }
  ^
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/lib/stan_math/stan/math/prim/fun/interp1.hpp:8:55: note: candidate function [with T1 = double, T2 = double, T3 = stan::math::var_value<double>]
        inline std::vector<return_type_t<T1, T2, T3>> interp1(const std::vector<T1>& xData, const std::vector<T2>& yData, const std::vector<T3>& xTest, std::ostream* pstream__ = nullptr)
                                                      ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:72:1: note: candidate function [with T0__ = double, T1__ = double, T2__ = stan::math::var_value<double>]
interp1(const std::vector<T0__>& x, const std::vector<T1__>& y,
^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:182:15: error: call to 'interp1' is ambiguous
      y_int = interp1(x, y, x_int, pstream__)[(1 - 1)];
              ^~~~~~~
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:392:14: note: in instantiation of function template specialization 'interp1_model_namespace::interp1_model::log_prob<true, true, stan::math::var_value<double> >' requested here
      return log_prob<propto__,jacobian__,T_>(vec_params_r, vec_params_i, pstream);
             ^
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/src/stan/model/model_base_crtp.hpp:129:50: note: in instantiation of function template specialization 'interp1_model_namespace::interp1_model::log_prob<true, true, stan::math::var_value<double> >' requested here
    return static_cast<const M*>(this)->template log_prob<true, true>(theta,
                                                 ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:82:3: note: in instantiation of member function 'stan::model::model_base_crtp<interp1_model_namespace::interp1_model>::log_prob_propto_jacobian' requested here
  ~interp1_model() final { }
  ^
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/lib/stan_math/stan/math/prim/fun/interp1.hpp:8:55: note: candidate function [with T1 = double, T2 = double, T3 = stan::math::var_value<double>]
        inline std::vector<return_type_t<T1, T2, T3>> interp1(const std::vector<T1>& xData, const std::vector<T2>& yData, const std::vector<T3>& xTest, std::ostream* pstream__ = nullptr)
                                                      ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:72:1: note: candidate function [with T0__ = double, T1__ = double, T2__ = stan::math::var_value<double>]
interp1(const std::vector<T0__>& x, const std::vector<T1__>& y,
^
In file included from /var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:1:
In file included from stan/src/stan/model/model_header.hpp:13:
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/src/stan/model/model_base_crtp.hpp:152:50: error: no matching member function for call to 'log_prob'
    return static_cast<const M*>(this)->template log_prob<false, false>(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:82:3: note: in instantiation of member function 'stan::model::model_base_crtp<interp1_model_namespace::interp1_model>::log_prob' requested here
  ~interp1_model() final { }
  ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:141:14: note: candidate template ignored: substitution failure [with propto__ = false, jacobian__ = false, T__ = stan::math::var_value<double>]
  inline T__ log_prob(std::vector<T__>& params_r__,
             ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:385:15: note: candidate function template not viable: requires at most 2 arguments, but 3 were provided
    inline T_ log_prob(Eigen::Matrix<T_,Eigen::Dynamic,1>& params_r,
              ^
In file included from /var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:1:
In file included from stan/src/stan/model/model_header.hpp:13:
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/src/stan/model/model_base_crtp.hpp:165:50: error: no matching member function for call to 'log_prob'
    return static_cast<const M*>(this)->template log_prob<false, true>(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:82:3: note: in instantiation of member function 'stan::model::model_base_crtp<interp1_model_namespace::interp1_model>::log_prob_jacobian' requested here
  ~interp1_model() final { }
  ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:141:14: note: candidate template ignored: substitution failure [with propto__ = false, jacobian__ = true, T__ = stan::math::var_value<double>]
  inline T__ log_prob(std::vector<T__>& params_r__,
             ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:385:15: note: candidate function template not viable: requires at most 2 arguments, but 3 were provided
    inline T_ log_prob(Eigen::Matrix<T_,Eigen::Dynamic,1>& params_r,
              ^
In file included from /var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:1:
In file included from stan/src/stan/model/model_header.hpp:13:
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/src/stan/model/model_base_crtp.hpp:178:50: error: no matching member function for call to 'log_prob'
    return static_cast<const M*>(this)->template log_prob<true, false>(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:82:3: note: in instantiation of member function 'stan::model::model_base_crtp<interp1_model_namespace::interp1_model>::log_prob_propto' requested here
  ~interp1_model() final { }
  ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:141:14: note: candidate template ignored: substitution failure [with propto__ = true, jacobian__ = false, T__ = stan::math::var_value<double>]
  inline T__ log_prob(std::vector<T__>& params_r__,
             ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:385:15: note: candidate function template not viable: requires at most 2 arguments, but 3 were provided
    inline T_ log_prob(Eigen::Matrix<T_,Eigen::Dynamic,1>& params_r,
              ^
In file included from /var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:1:
In file included from stan/src/stan/model/model_header.hpp:13:
/Users/hyunjimoon/Dropbox/stan/customFunc/customFunc_cmdstan/stan/src/stan/model/model_base_crtp.hpp:191:50: error: no matching member function for call to 'log_prob'
    return static_cast<const M*>(this)->template log_prob<true, true>(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:82:3: note: in instantiation of member function 'stan::model::model_base_crtp<interp1_model_namespace::interp1_model>::log_prob_propto_jacobian' requested here
  ~interp1_model() final { }
  ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:141:14: note: candidate template ignored: substitution failure [with propto__ = true, jacobian__ = true, T__ = stan::math::var_value<double>]
  inline T__ log_prob(std::vector<T__>& params_r__,
             ^
/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f.hpp:385:15: note: candidate function template not viable: requires at most 2 arguments, but 3 were provided
    inline T_ log_prob(Eigen::Matrix<T_,Eigen::Dynamic,1>& params_r,
              ^
8 errors generated.
make: *** [/var/folders/f0/xzz_g2mx5qb8982lmv5jylgc0000gn/T/Rtmp0oe5fP/model-11c252c3b3d4f] Error 1
Error: An error occured during compilation! See the message above for more information.
```
